notify/templates.go

```diff
@@ -0,0 +1,240 @@
+	Kind TemplateErrorKind `json:"kind"`
+
+	// Error message.
+	Message string `json:"message"`
+}
+
+type TemplateErrorKind string
+
+const (
+	InvalidTemplate TemplateErrorKind = "invalid_template"
+	ExecutionError  TemplateErrorKind = "execution_error"
+)
+
+// TestTemplate tests the given template string against the given alerts. Existing templates are used to provide context for the test.
+// If an existing template of the same filename as the one being tested is found, it will not be used as context.
+func (am *GrafanaAlertmanager) TestTemplate(ctx context.Context, c TestTemplatesConfigBodyParams) (*TestTemplatesResults, error) {
+	definitions, err := parseTestTemplate(c.Name, c.Template)
+	if err != nil {
+		return &TestTemplatesResults{
+			Errors: []TestTemplatesErrorResult{{
+				Kind:    InvalidTemplate,
+				Message: err.Error(),
+			}},
+		}, nil
+	}
+
+	// Recreate the current template without the definition blocks that are being tested. This is so that any blocks that were removed don't get defined.
+	paths := make([]string, 0)
+	for _, name := range am.templates {
+		if name == c.Name {
+			// Skip the existing template of the same name as we're going to parse the one for testing instead.
+			continue
+		}
+		paths = append(paths, filepath.Join(am.workingDirectory, name))
+	}
+
+	// Parse current templates.
+	var newTextTmpl *tmpltext.Template
+	var captureTemplate template.Option = func(text *tmpltext.Template, _ *tmplhtml.Template) {
+		newTextTmpl = text
+	}
+	newTmpl, err := templates.FromGlobs(paths, captureTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse test template.
+	_, err = newTextTmpl.Parse(c.Template)
+	if err != nil {
+		// This shouldn't happen since we already parsed the template above.
+		return nil, err
+	}
+
+	externalURL, err := url.Parse(am.ExternalURL())
+	if err != nil {
+		return nil, err
+	}
+	newTmpl.ExternalURL = externalURL
+
+	// Prepare the context.
+	alerts := v2.OpenAPIAlertsToAlerts(c.Alerts)
+	ctx = notify.WithReceiverName(ctx, "test receiver")
+	ctx = notify.WithGroupLabels(ctx, model.LabelSet{"group_label": "group_label_value"})
+
+	var tmplErr error
+	templater, _ := templates.TmplText(ctx, newTmpl, alerts, am.logger, &tmplErr)
+
+	// Iterate over each definition in the template and evaluate it.
+	var results TestTemplatesResults
+	for _, def := range definitions {
+		s := fmt.Sprintf(`{{ template "%s" . }}`, def)
+		val := templater(s)
+		if tmplErr != nil {
+			results.Errors = append(results.Errors, TestTemplatesErrorResult{
+				Name:    def,
+				Kind:    ExecutionError,
+				Message: tmplErr.Error(),
+			})
+			tmplErr = nil
+			continue
+		}
+
+		results.Results = append(results.Results, TestTemplatesResult{
+			Name: def,
+			Text: val,
+		})
+	}
+
+	return &results, nil
+}
+
+// parseTestTemplate parses the test template and returns the top-level definitions that should be interpolated as results.
+func parseTestTemplate(name string, template string) ([]string, error) {
```

Comment:
I think this and functions below should live in `templates` package because it's too specific for template processing